examples/pipelines/fl_cross_silo_factory/fl_factory.py

```diff
@@ -0,0 +1,503 @@
+        self.silos.append(
+            {
+                "compute": compute,
+                "datastore": datastore,
+                "custom_input_args": custom_input_args or {},
+            }
+        )
+
+    def custom_fl_data_output(
+        self, datastore_name, output_name, unique_id="${{name}}", round=None
+    ):
+        """Produces a path to store the data during FL training.
+
+        Args:
+            datastore_name (str): name of the Azure ML datastore
+            output_name (str): a name unique to this output
+            unique_id (str): a unique id for the run (default: inject run id with ${{name}})
+            round (str): an round id if relevant
+
+        Returns:
+            data_path (str): direct url to the data path to store the data
+        """
+        data_path = f"azureml://datastores/{datastore_name}/paths/federated_learning/{output_name}/{unique_id}/"
+        if round:
+            data_path += f"round_{round}/"
+
+        return Output(type=AssetTypes.URI_FOLDER, mode="mount", path=data_path)
+
+    def getUniqueIdentifier(self, length=8):
+        """Generates a random string and concatenates it with today's date
+
+        Args:
+            length (int): length of the random string (default: 8)
+
+        """
+        str = string.ascii_lowercase
+        date = datetime.date.today().strftime("%Y_%m_%d_")
+        return date + "".join(random.choice(str) for i in range(length))
+
+    def anchor_step_in_silo(
+        self,
+        pipeline_step,
+        compute,
+        output_datastore,
+        model_output_datastore=None,
+        tags={},
+        description=None,
+    ):
+        """Takes a step and enforces the right compute/datastore config"""
+        # make sure the compute corresponds to the silo
+        pipeline_step.compute = compute
+
+        # make sure every output data is written in the right datastore
+        for key in pipeline_step.outputs:
+            _output = getattr(pipeline_step.outputs, key)
+            if _output.type == AssetTypes.CUSTOM_MODEL:
+                setattr(
+                    pipeline_step.outputs,
+                    key,
+                    self.custom_fl_data_output(
+                        model_output_datastore or output_datastore, key
+                    ),
+                )
+            else:
+                setattr(
+                    pipeline_step.outputs,
+                    key,
+                    self.custom_fl_data_output(output_datastore, key),
+                )
+
+        return pipeline_step
+
+    def build_basic_fl_pipeline(
+        self,
+        silo_preprocessing,
+        silo_training,
+        orchestrator_aggregation,
+        iterations=1,
+        **training_kwargs,
+    ):
+        @pipeline(
+            description=f'FL cross-silo basic pipeline and the unique identifier is "{self.unique_identifier}" that can help you to track files in the storage account.',
+        )
```

Comment:
preprocessing step at silo better to be optional in fl-factory logics, and fl-factory logics need to handle the scenario whether or not users provide/use "preprocessing component" step. Refer to details for https://github.com/Azure/shrike/blob/main/shrike/pipeline/federated_learning.py, where preprocessing and post-processing step are optional, only silo-train and server-agg are hard requirement for FL API.

It is common for scenarios where customer silos don't need any extra preprocessing on the silo's data (e.g., already cleaned, or the silo's data needs offline heterogeneous preprocessing, e.g., different languages data from different companies, and can't use a shared common preprocessing step here.)

Reply:
We could make each of those steps an optional argument in the builder, I'll refactor in this direction.